### ISSUE TITLE
twisted.web.static no longer works with HTTP pipelining

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3919]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ivank reported|
|-|-|
|Trac ID|trac#3919|
|Type|release blocker: regression|
|Created|2009-07-18 17:10:40Z|
|Branch|https://github.com/twisted/twisted/tree/dont-pause-a-pull-producer-3919-2|

I used twisted.web.static.File to serve a directory (Twisted trunk@27161). With pipelining enabled in Firefox 3.5, the server showed this traceback very often:

```

2009-07-18 06:30:34-0700 [HTTPChannel,1,192.168.0.3] Unhandled Error
        Traceback (most recent call last):
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/protocols/basic.py", line 238, in dataReceived
            why = self.lineReceived(line)
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/web/http.py", line 1455, in lineReceived
            self.allContentReceived()
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/web/http.py", line 1523, in allContentReceived
            req.requestReceived(command, path, version)
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/web/http.py", line 725, in requestReceived
            self.process()
        --- <exception caught here> ---
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/web/server.py", line 126, in process
            self.render(resrc)
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/web/server.py", line 133, in render
            body = resrc.render(self)
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/web/static.py", line 623, in render
            producer.start()
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/web/static.py", line 706, in start
            self.request.registerProducer(self, False)
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/web/http.py", line 750, in registerProducer
            producer.pauseProducing()
        exceptions.AttributeError: 'NoRangeStaticProducer' object has no attribute 'pauseProducing'

[...]

2009-07-18 06:30:35-0700 [HTTPChannel,5,192.168.0.3] Unhandled Error
        Traceback (most recent call last):
        Failure: exceptions.RuntimeError: Producer was not unregistered for /url1

2009-07-18 06:30:35-0700 [HTTPChannel,5,192.168.0.3] Unhandled Error
        Traceback (most recent call last):
        Failure: exceptions.RuntimeError: Producer was not unregistered for /url2

[...]

```

<details><summary>Searchable metadata</summary>

```
trac-id__3919 3919
type__release_blocker__regression release blocker: regression
reporter__ivank ivank
priority__high high
milestone__Twisted_9_0 Twisted-9.0
branch__branches_dont_pause_a_pull_producer_3919_2 branches/dont-pause-a-pull-producer-3919-2
branch_author__mwh mwh
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1247937040000000 1247937040000000
changetime__1248989835000000 1248989835000000
version__None None
owner__ 
cc__mwh
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|ivank commented|
|-|-|

An interactive test case:

```python

from twisted.web import server, static
from twisted.internet import reactor
from twisted.python import log, filepath

import sys, os
log.startLogging(sys.stdout)

files = filepath.FilePath('.').temporarySibling()
files.makedirs()

index = 'Making a lot of requests. Now check for tracebacks on the server.<br>'

for i in xrange(20):
        files.child("%d.txt" % (i,)).setContent('//'*30000)
        index += '<script src="%d.txt"></script>' % (i,)

files.child("index.html").setContent(index)

reactor.listenTCP(49999, server.Site(static.File(files.path)), interface='0')

reactor.run()

```

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|ivank commented|
|-|-|

I cannot reproduce the problem from before t.w.static.File-multiple-ranges-3574-2 was merged.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_5"></a>|@mwhudson commented|
|-|-|

I guess a "def pauseProducing(self): pass" will make this go away (it's what t.w.static.FileTransfer does) but as exarkun said on IRC, it's wrong that pauseProducing is called on an IPullProducer.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_6"></a>|@mwhudson commented|
|-|-|

http://pastebin.ubuntu.com/221658/ is a rather brainless fix.

This area is not at all well tested, it seems.

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|ivank commented|
|-|-|

I also see this traceback on occasion. I don't know exactly what caused it.

```
2009-07-19 23:24:17-0700 [HTTPChannel,16,192.168.0.3] Unhandled Error
        Traceback (most recent call last):
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/application/app.py", line 338, in runReactorWithLogging
            reactor.run()
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/internet/base.py", line 1168, in run
            self.mainLoop()
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/internet/base.py", line 1180, in mainLoop
            self.doIteration(t)
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/internet/selectreactor.py", line 142, in doSelect
            _logrun(selectable, _drdw, selectable, method, dict)
        --- <exception caught here> ---
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/python/log.py", line 86, in callWithLogger
            return callWithContext({"system": lp}, func, *args, **kw)
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/python/log.py", line 71, in callWithContext
            return context.call({ILogContext: newCtx}, func, *args, **kw)
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/python/context.py", line 59, in callWithContext
            return self.currentContext().callWithContext(ctx, func, *args, **kw)
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/python/context.py", line 37, in callWithContext
            return func(*args,**kw)
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/internet/selectreactor.py", line 158, in _doReadOrWrite
            self._disconnectSelectable(selectable, why, method=="doRead")
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/internet/posixbase.py", line 192, in _disconnectSelectable
            selectable.connectionLost(f)
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/internet/tcp.py", line 524, in connectionLost
            abstract.FileDescriptor.connectionLost(self, reason)
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/internet/abstract.py", line 66, in connectionLost
            self.producer.stopProducing()
        exceptions.AttributeError: 'NoRangeStaticProducer' object has no attribute 'stopProducing'
```

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|ivank commented|
|-|-|

Sorry, above traceback was with a modified Twisted. It happens with trunk too:

```
2009-07-19 23:31:59-0700 [HTTPChannel,14,192.168.0.3] Unhandled Error
        Traceback (most recent call last):
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/application/app.py", line 348, in runReactorWithLogging
            reactor.run()
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/internet/base.py", line 1166, in run
            self.mainLoop()
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/internet/base.py", line 1178, in mainLoop
            self.doIteration(t)
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/internet/selectreactor.py", line 140, in doSelect
            _logrun(selectable, _drdw, selectable, method, dict)
        --- <exception caught here> ---
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/python/log.py", line 84, in callWithLogger
            return callWithContext({"system": lp}, func, *args, **kw)
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/python/log.py", line 69, in callWithContext
            return context.call({ILogContext: newCtx}, func, *args, **kw)
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/python/context.py", line 59, in callWithContext
            return self.currentContext().callWithContext(ctx, func, *args, **kw)
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/python/context.py", line 37, in callWithContext
            return func(*args,**kw)
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/internet/selectreactor.py", line 156, in _doReadOrWrite
            self._disconnectSelectable(selectable, why, method=="doRead")
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/internet/posixbase.py", line 191, in _disconnectSelectable
            selectable.connectionLost(f)
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/internet/tcp.py", line 513, in connectionLost
            abstract.FileDescriptor.connectionLost(self, reason)
          File "/opt/Python-latest/lib/python2.7/site-packages/twisted/internet/abstract.py", line 64, in connectionLost
            self.producer.stopProducing()
        exceptions.AttributeError: 'NoRangeStaticProducer' object has no attribute 'stopProducing'
```

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|ivank commented|
|-|-|

This is my current fix:

```
        if self.queued:
            if streaming:
                producer.pauseProducing()
        else:
            self.transport.registerProducer(producer, streaming)
```

And this method added to `class StaticProducer(object):`:

```
    def stopProducing(self):
        self.fileObject.close()
        self.request = None
```

This seems to fix the problem. Note that `if streaming and self.queued:` is erroneous.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_11"></a>|@mwhudson commented|
|-|-|

Those fixes make sense.

Now, we need test cases :)

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_12"></a>|@mwhudson commented|
|-|-|

I decided to stop punishing myself with Subversion and have a fix for this in a Bazaar branch: https://code.edge.launchpad.net/~mwhudson/twisted/dont-pause-a-pull-producer-3919/+merge/9285

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)|@mwhudson removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @mwhudson|
|-|-|

The fix basically looks right to me, and the greatly increased test coverage is nice.

  1. There are some coding standard issues:
    1. Test methods should be named like `test_fooBar`
    1. Methods should be separated from each other by two blank lines.  Many of the methods related to producing a request body in static.py are only separated by one blank line.
    1. Docstring convention is to not do `"""Foo` but instead `"""\nFoo`. (`StaticProducer.stopProducing`)
  1. `NoRangeStaticProducerTests` doesn't define any test methods.
  1. `SpyProducer` is almost an identical re-implementation of `twisted.test.test_internet.DummyProducer`.
  1. `twisted.web.static.StaticProducer` should have `@ivar` docs (probably replacing the `@param` docs of its `__init__`)
  1. `test_registerProducer_when_queued_doesnt_register_push_producer` could probably also assert that `http.StringTransport` has no `registerProducer` method, so it will break automatically when its assumption is broken.
  1. `assertIs` should be `assertIdentical` (or it'll break on older versions of Python)

It'd be good to see a run of this code on buildbot, on all the supported platforms.  Also, what's the plan for getting the code into svn?

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_15"></a>|@mwhudson removed owner|
|-|-|

Thanks for the review!

I think I've addressed all your points.

Here's the update diff against trunk: http://pastebin.ubuntu.com/234150/

And here's the diff from what you reviewed: http://pastebin.ubuntu.com/234151/

To get this into subversion, I guess I'll push the branch using bzr-svn, and then test with buildbot.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_17"></a>|@mwhudson commented|
|-|-|

OK, code jammed into svn it seems.

Is there some magic way to run a branch on all supported platforms in buildbot magically, or do you have to click "force build" a lot?

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_18"></a>|@mwhudson commented|
|-|-|

Thanks to therve's pointer, results are here: http://buildbot.twistedmatrix.com/boxes-supported?branch=branches/dont-pause-a-pull-producer-3919

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

(In [27207]) Branching to 'dont-pause-a-pull-producer-3919-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun commented|
|-|-|

(In [27208]) Merge all the changes from the previous branch, including the ones Combinator doesn't see

refs [#3919](https://github.com/twisted/twisted/issues/3919)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun set owner to @mwhudson|
|-|-|

The docstrings of `test_registerProducerWhenQueuedPausesPushProducer` and `test_registerProducerWhenQueuedDoesntPausePullProducer` each talk about the wrong kind of producer.  The former talks about `IPullProducer`, the latter about `IPushProducer`.

`test_registerProducerWhenNotQueuedRegistersPushProducer` and `test_registerProducerWhenNotQueuedRegistersPullProducer` each pass the wrong value for the 2nd argument to the `registerProducer` call they make.  The former passes false, indicating a pull (non-streaming) producer; the latter passes true, indicating a push (streaming) producer.

There is a problem with the way the svn branch was constructed.  The initial revision in which it exists also includes changes to `test_static.py`.  This violates an assumption made by Combinator, and so Combinator is not able to entirely merge the branch.  I've created a new branch which follows the normal svn workflow.  It would be nice to have this changed in bzr-svn.  Until then, I don't think anyone should use bzr-svn for Twisted development (I had to go quite out of my way to notice this problem; next time, the reviewer might not notice).

Otherwise, looks good to me.  Please merge (the new branch (with Combinator)) when the above two issues are addressed.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_22"></a>|@mwhudson set status to `closed`|
|-|-|

(In [27210]) Merge dont-pause-a-pull-producer-3919-2.

Author: mwh
Reviewer: exarkun
Fixes: [#3919](https://github.com/twisted/twisted/issues/3919)

Fix two problems caused by the recent changes to serving static files:

1. Change twisted.web.http.Request to better respect the interface of
   its producer.

2. Change twisted.web.static.StaticProducer (and subclasses) to
   properly implement IPullProducer.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_23"></a>|@mwhudson set status to `reopened`|
|-|-|

(In [27211]) Revert r27210.

I'm an idiot, and hadn't run the right tests before merging.

Reopens [#3919](https://github.com/twisted/twisted/issues/3919).

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_24"></a>|@mwhudson set status to `closed`|
|-|-|

(In [27213]) Merge dont-pause-a-pull-producer-3919-2.

Author: mwh
Reviewer: exarkun
Fixes: [#3919](https://github.com/twisted/twisted/issues/3919)

Fix two problems caused by the recent changes to serving static files,
having run all the twisted.web tests this time:

1. Change twisted.web.http.Request to better respect the interface of
   its producer.

2. Change twisted.web.static.StaticProducer (and subclasses) to
   properly implement IPullProducer.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
